### PR TITLE
[JENKINS-61683] Initial implementation of GitLab Hook parameters as env variable

### PR DESCRIPTION
**Pull request by baymac**
I am trying add GitLab hook parameters as env variable to builds. I get this error when there is an event hook received by Jenkins:

```
Push event to branch master
Started by GitLab push by Parichay Barpanda
Querying the current revision of branch master...
Current revision of branch master is 182d9bb541ac58ed0867bbd5bc9fd8df45ec68ec
Obtained Jenkinsfile from 182d9bb541ac58ed0867bbd5bc9fd8df45ec68ec
Running in Durability level: MAX_SURVIVABILITY
[GitLab Pipeline Status] Notifying branch build status: FAILED test/master: There was a failure building this commit
[GitLab Pipeline Status] Notified
java.lang.UnsupportedOperationException: Refusing to marshal org.gitlab4j.api.webhook.PushEvent for security reasons; see https://jenkins.io/redirect/class-filter/
	at hudson.util.XStream2$BlacklistedTypesConverter.marshal(XStream2.java:542)
	at com.thoughtworks.xstream.core.AbstractReferenceMarshaller.convert(AbstractReferenceMarshaller.java:69)
	at com.thoughtworks.xstream.core.TreeMarshaller.convertAnother(TreeMarshaller.java:58)
	at com.thoughtworks.xstream.core.AbstractReferenceMarshaller$1.convertAnother(AbstractReferenceMarshaller.java:84)
	at hudson.util.RobustReflectionConverter.marshallField(RobustReflectionConverter.java:263)
	at hudson.util.RobustReflectionConverter$2.writeField(RobustReflectionConverter.java:250)
Caused: java.lang.RuntimeException: Failed to serialize io.jenkins.plugins.gitlabbranchsource.Cause.GitLabPushCauseData#pushEvent for class io.jenkins.plugins.gitlabbranchsource.Cause.GitLabPushCauseData
	at hudson.util.RobustReflectionConverter$2.writeField(RobustReflectionConverter.java:254)
	at hudson.util.RobustReflectionConverter$2.visit(RobustReflectionConverter.java:222)
	at com.thoughtworks.xstream.converters.reflection.PureJavaReflectionProvider.visitSerializableFields(PureJavaReflectionProvider.java:138)
	at hudson.util.RobustReflectionConverter.doMarshal(RobustReflectionConverter.java:208)
	at hudson.util.RobustReflectionConverter.marshal(RobustReflectionConverter.java:149)
	at com.thoughtworks.xstream.core.AbstractReferenceMarshaller.convert(AbstractReferenceMarshaller.java:69)
	at com.thoughtworks.xstream.core.TreeMarshaller.convertAnother(TreeMarshaller.java:58)
	at com.thoughtworks.xstream.core.AbstractReferenceMarshaller$1.convertAnother(AbstractReferenceMarshaller.java:84)
	at hudson.util.RobustReflectionConverter.marshallField(RobustReflectionConverter.java:263)
	at hudson.util.RobustReflectionConverter$2.writeField(RobustReflectionConverter.java:250)
Caused: java.lang.RuntimeException: Failed to serialize io.jenkins.plugins.gitlabbranchsource.GitLabWebHookCause#gitLabPushCauseData for class io.jenkins.plugins.gitlabbranchsource.GitLabWebHookCause
	at hudson.util.RobustReflectionConverter$2.writeField(RobustReflectionConverter.java:254)
	at hudson.util.RobustReflectionConverter$2.visit(RobustReflectionConverter.java:222)
	at com.thoughtworks.xstream.converters.reflection.PureJavaReflectionProvider.visitSerializableFields(PureJavaReflectionProvider.java:138)
	at hudson.util.RobustReflectionConverter.doMarshal(RobustReflectionConverter.java:208)
	at hudson.util.RobustReflectionConverter.marshal(RobustReflectionConverter.java:149)
	at com.thoughtworks.xstream.core.AbstractReferenceMarshaller.convert(AbstractReferenceMarshaller.java:69)
	at com.thoughtworks.xstream.core.TreeMarshaller.convertAnother(TreeMarshaller.java:58)
	at com.thoughtworks.xstream.core.TreeMarshaller.convertAnother(TreeMarshaller.java:43)
	at com.thoughtworks.xstream.core.AbstractReferenceMarshaller$1.convertAnother(AbstractReferenceMarshaller.java:88)
	at com.thoughtworks.xstream.converters.collections.AbstractCollectionConverter.writeItem(AbstractCollectionConverter.java:64)
	at com.thoughtworks.xstream.converters.collections.MapConverter.marshal(MapConverter.java:78)
	at com.thoughtworks.xstream.core.AbstractReferenceMarshaller.convert(AbstractReferenceMarshaller.java:69)
	at com.thoughtworks.xstream.core.TreeMarshaller.convertAnother(TreeMarshaller.java:58)
	at com.thoughtworks.xstream.core.AbstractReferenceMarshaller$1.convertAnother(AbstractReferenceMarshaller.java:84)
	at hudson.util.RobustReflectionConverter.marshallField(RobustReflectionConverter.java:263)
	at hudson.util.RobustReflectionConverter$2.writeField(RobustReflectionConverter.java:250)
Caused: java.lang.RuntimeException: Failed to serialize hudson.model.CauseAction#causeBag for class hudson.model.CauseAction
	at hudson.util.RobustReflectionConverter$2.writeField(RobustReflectionConverter.java:254)
	at hudson.util.RobustReflectionConverter$2.visit(RobustReflectionConverter.java:222)
	at com.thoughtworks.xstream.converters.reflection.PureJavaReflectionProvider.visitSerializableFields(PureJavaReflectionProvider.java:138)
	at hudson.util.RobustReflectionConverter.doMarshal(RobustReflectionConverter.java:208)
	at hudson.util.RobustReflectionConverter.marshal(RobustReflectionConverter.java:149)
	at hudson.util.XStream2$PassthruConverter.marshal(XStream2.java:490)
	at hudson.util.XStream2$AssociatedConverterImpl.marshal(XStream2.java:460)
	at com.thoughtworks.xstream.core.AbstractReferenceMarshaller.convert(AbstractReferenceMarshaller.java:69)
	at com.thoughtworks.xstream.core.TreeMarshaller.convertAnother(TreeMarshaller.java:58)
	at com.thoughtworks.xstream.core.TreeMarshaller.convertAnother(TreeMarshaller.java:43)
	at com.thoughtworks.xstream.core.AbstractReferenceMarshaller$1.convertAnother(AbstractReferenceMarshaller.java:88)
	at com.thoughtworks.xstream.converters.collections.AbstractCollectionConverter.writeItem(AbstractCollectionConverter.java:64)
	at com.thoughtworks.xstream.converters.collections.CollectionConverter.marshal(CollectionConverter.java:74)
	at com.thoughtworks.xstream.core.AbstractReferenceMarshaller.convert(AbstractReferenceMarshaller.java:69)
	at com.thoughtworks.xstream.core.TreeMarshaller.convertAnother(TreeMarshaller.java:58)
	at com.thoughtworks.xstream.core.AbstractReferenceMarshaller$1.convertAnother(AbstractReferenceMarshaller.java:84)
	at hudson.util.RobustReflectionConverter.marshallField(RobustReflectionConverter.java:263)
	at hudson.util.RobustReflectionConverter$2.writeField(RobustReflectionConverter.java:250)
Caused: java.lang.RuntimeException: Failed to serialize hudson.model.Actionable#actions for class org.jenkinsci.plugins.workflow.job.WorkflowRun
	at hudson.util.RobustReflectionConverter$2.writeField(RobustReflectionConverter.java:254)
	at hudson.util.RobustReflectionConverter$2.visit(RobustReflectionConverter.java:222)
	at com.thoughtworks.xstream.converters.reflection.PureJavaReflectionProvider.visitSerializableFields(PureJavaReflectionProvider.java:138)
	at hudson.util.RobustReflectionConverter.doMarshal(RobustReflectionConverter.java:208)
	at hudson.util.RobustReflectionConverter.marshal(RobustReflectionConverter.java:149)
	at com.thoughtworks.xstream.core.AbstractReferenceMarshaller.convert(AbstractReferenceMarshaller.java:69)
	at com.thoughtworks.xstream.core.TreeMarshaller.convertAnother(TreeMarshaller.java:58)
	at com.thoughtworks.xstream.core.TreeMarshaller.convertAnother(TreeMarshaller.java:43)
	at com.thoughtworks.xstream.core.TreeMarshaller.start(TreeMarshaller.java:82)
	at com.thoughtworks.xstream.core.AbstractTreeMarshallingStrategy.marshal(AbstractTreeMarshallingStrategy.java:37)
	at com.thoughtworks.xstream.XStream.marshal(XStream.java:1026)
	at com.thoughtworks.xstream.XStream.marshal(XStream.java:1015)
	at com.thoughtworks.xstream.XStream.toXML(XStream.java:988)
	at hudson.XmlFile.write(XmlFile.java:193)
Caused: java.io.IOException
	at hudson.XmlFile.write(XmlFile.java:200)
	at org.jenkinsci.plugins.workflow.support.PipelineIOUtils.writeByXStream(PipelineIOUtils.java:30)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun.save(WorkflowRun.java:1191)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun.run(WorkflowRun.java:317)
	at hudson.model.ResourceController.execute(ResourceController.java:97)
	at hudson.model.Executor.run(Executor.java:429)
Finished: FAILURE
```

The Jenkinsfile contains:

```
node {
  echo sh(returnStdout: true, script: 'env')
}
```

I think the problem is with serialising the Cause Data (i.e. push Event, MR event etc.). How can I serialise the pushEvent class so that Jenkins will be able to save it?

Update: Removed event objects from causedataobjects so need not be serialized.

### CODE REVIEW

**Comment by baymac** [comment]
This tiring to map all variables key to their value in environment variables, any library that can solve this?

**Comment by baymac** [comment]
Tadaa! Hook urls can be seen in the builds :)
Screenshot 2020-04-25 at 2 28 14 PM

**Comment by baymac** [comment]
Never had I written so much boiler plate code 😓 😓 😓

File: src/main/java/io/jenkins/plugins/gitlabbranchsource/Cause/GitLabMergeRequestCauseData.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+import static io.jenkins.plugins.gitlabbranchsource.Cause.GitLabCauseUtils.defaultListSize;
+import static io.jenkins.plugins.gitlabbranchsource.Cause.GitLabCauseUtils.defaultVisibilityString;
+import static org.apache.commons.lang.StringUtils.defaultString;
+
+@ExportedBean
+public class GitLabMergeRequestCauseData {
+
+    private Map<String, String> variables = new HashMap<>();
+
+    public GitLabMergeRequestCauseData(MergeRequestEvent mergeRequestEvent) {
```
Comment (jetersen) [inline src/main/java/io/jenkins/plugins/gitlabbranchsource/Cause/GitLabMergeRequestCauseData.java:22]: The classes should share a common class to reduce the duplicated code.

**Comment by baymac** [comment]
@casz Is there a way to get cause data from a Job? E.g. In Branch API Plugin MultibranchProject Job type is used as filter https://github.com/jenkinsci/branch-api-plugin/blob/c4d394415cf25b6890855a08360119313f1330d2/src/main/java/jenkins/branch/BranchNameContributor.java. Also want to know what is the type of Build for MultibranchProject?

**Comment by baymac** [comment]
Is there any use case for users using this plugin and not using WorkflowRun type of Build? Otherwise in the environment contributor we can just add gitlab webhook variables for WorkflowRun type of builds.

Also I am trying to write test case for environment contributor and I want to reuse some of the test classes of gitlab4j but I am unable. How can I be able to import them?

**Comment by baymac** [comment]
I gave up on writing test, I feel I don't have right knowledge. Will add them later. Is it ready to be merged @jetersen?

**Comment by jetersen** [comment]
As both tag event and push event share the same abstract push you should be able to reduce the copy pasta 🗡️

Not sure whether it is feasible to extend the three classes and add a common interface to completely remove all the duplicated code.

**Comment by baymac** [comment]
> As both tag event and push event share the same abstract push you should be able to reduce the copy pasta 🗡️
> 
> Not sure whether it is feasible to extend the three classes and add a common interface to completely remove all the duplicated code.

Agreed but I will not be able to spent time on this for about a week. At the same time I want to release this feature along with next release (after merging #88).

**Comment by ababushk** [comment]
Hi @baymac @jetersen !
This is very useful feature, thanks for your work!
I want to try to implement retrieving of some data provided by this PR, but when build is started either manually or via branch indexing. We can access this data by querying GitLab API. Could you please advice me if it is possible in current plugin design?

**Comment by asoldo11** [comment]
Hi guys, is there any progress regarding this super useful feature?

**Pull request merged**